docs/advanced/classes.rst

```diff
@@ -1384,11 +1384,29 @@ You can do that using ``py::custom_type_setup``:
            auto *type = &heap_type->ht_type;
            type->tp_flags |= Py_TPFLAGS_HAVE_GC;
            type->tp_traverse = [](PyObject *self_base, visitproc visit, void *arg) {
+               // https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_traverse
+   #if PY_VERSION_HEX >= 0x03090000  // Python 3.9
+               Py_VISIT(Py_TYPE(self_base));
+   #endif
+               if (!py::detail::is_holder_constructed(self_base)) [[unlikely]] {
```

Comment:
This will look a lot more straightforward if we simply guard the existing code with

```
if (py::detail::is_holder_constructed(self_base)) {
```

Could you please update the code in tests/test_custom_type_setup.cpp accordingly?

Do we still need the new test?

Reply:
> Do we still need the new test?

I think it can serve as a backlog if we implement https://github.com/pybind/pybind11/pull/5654#discussion_r2088081105 eventually.

Reply:
I think what'll be most practical and best at this stage:

* Pull out the changes to docs/advanced/classes.rst and include/pybind11/detail/value_and_holder.h into a separate PR.

* Update tests/test_custom_type_setup.cpp to be 1:1 with the updated documentation in classes.rst

Merge that PR.

Then you can merge the new master here and work on this PR in the future.

The tests will need more work.

Reply:
Opened #5669.